.ci/Treecheck/Treecheck.jl

```diff
@@ -0,0 +1,187 @@
+            end
+        end
+    end
+
+    num_changed_packages = length(package_list)
+
+    if num_changed_packages > 10
+        # This step prevents us from launching too many CI jobs (e.g. a hundred jobs)
+
+        msg = "Too many changed packages: $(num_changed_packages)"
+        error(msg)
+    end
+
+    dict = Dict()
+    dict["include"] = []
+
+    for package in package_list
+        matrix_entry = Dict(
+            "PACKAGE_NAME" => package.name,
+            "PACKAGE_UUID" => package.uuid_str,
+        )
+        push!(dict["include"], matrix_entry)
+    end
+
+    # Necessary to avoid erroring on empty matrix:
+    if isempty(dict["include"])
+        json_str = ""
+    else
+        json_str = JSON3.write(dict)
+    end
+
+    println(Base.stderr, json_str)
+
+    gha_set_output("matrix", json_str)
+
+    return json_str
+end
+
+
+
+function _treehash_to_commithash(treehash::AbstractString)
+    cmd_1 = `git log --pretty=raw`
+    cmd_2 = `grep -B 1 "$(treehash)"`
+    my_pipeline = pipeline(cmd_1, cmd_2)
+    str = strip(_tryread(my_pipeline))
+    r = r"commit ([A-Za-z0-9]*?)\n"
+    m = match(r, str)
+    commithash = strip(m[1])
+    return commithash
+end
+
+function _tryread(cmd::Base.AbstractCmd)
+    str = try
+        read(cmd, String)
+    catch ex
+        # If an exception is thrown, run the command again, but send stdout/stderr to the CI log
+        run(cmd)
+
+        rethrow()
+    end
+    return str
+end
+
+check(package_uuid_str::AbstractString) = check(RegistryToml(), package_uuid_str)
+function check(registrytoml::RegistryToml, package_uuid_str::AbstractString)
+    package_name = registrytoml.dict["packages"][package_uuid_str]["name"]
+    package_relpath = registrytoml.dict["packages"][package_uuid_str]["path"]
+    package_abspath = joinpath(registrytoml.root, package_relpath)
+    package_toml = joinpath(package_abspath, "Package.toml")
+    versions_toml = joinpath(package_abspath, "Versions.toml")
+    package_dict = TOML.parsefile(package_toml)
+    versions_dict = TOML.parsefile(versions_toml)
+    package_git_repo_url = package_dict["repo"]
+    @testset "Treecheck for package: $(package_name)" begin
+        @test !isempty(versions_dict)
+        mktempdir() do tmpdir
+            cd(tmpdir) do
+                run(`git clone "$(package_git_repo_url)" MYCLONEDIR`)
+                cd("MYCLONEDIR") do
+                    gitrepo_libgit2 = LibGit2.GitRepo(".")
+
+                    # For each tree, make sure the tree exists
+                    @testset for (k, v) in pairs(versions_dict)
+                        treehash = v["git-tree-sha1"]
+                        tree_libgit2 = LibGit2.GitTree(gitrepo_libgit2, LibGit2.GitHash(treehash))
+                        @test tree_libgit2 isa LibGit2.GitTree
+                    end
+
+                    treehash_to_commithash = Dict{String, String}()
+
+                    # For each tree, make sure that at least one commit points to that tree
```

Comment:
There are a couple problems here.
1. Only commits that can be traced from HEAD are detected. That might be too restrictive.
2. There is no support for subdir packages.
3. For performance it would be better to do the `git log` once and cache the result.

I'm also unsure why we are looking for commit hashes in the first place.

Reply:
Ahhhh, you are right, the commit stuff doesn't work for subdir packages.

In theory we can, for each commit, navigate to the subdir and then check that tree, but the problem is that the registry only stores one subdir, but the true subdir might be different for each commit.

I'm going to drop all of the commit stuff.

Reply:
I've removed the commit stuff in https://github.com/JuliaRegistries/General/pull/136859/commits/3209879f8014c2d606ec3dfedb57b4df67b77673